### PR TITLE
Autotuning

### PR DESCRIPTION
**Pull request by Yanzhiwen001**
This version can support Scallop and StringTie with input tunable parameters consisting of both categorical and continuous types, but not working when you only have continuous parameters as input.

### CODE REVIEW

File: BO_training/bo/localbo_utils.py [hunk 1]
```diff
@@ -407,14 +407,22 @@ def interleaved_search(x_center, f: Callable,
...
     # x0_cat = np.array([deepcopy(x_center[cat_dims]) for _ in range(n_restart)])
+    #for continuos parameters
     seed = np.random.randint(int(1e6))
     sobol = SobolEngine(len(cont_dims), scramble=True, seed=seed)
     x0_cont = sobol.draw(n_restart).cpu().detach().numpy()
     x0_cont = lb + (ub - lb) * x0_cont
-    x0 = np.hstack((x0_cat, x0_cont))
+    #add category parameters
+    #fix case for no catagory data
+    if (len(cat_dims) == 0):
```
Comment (gmarcais) [inline BO_training/bo/localbo_utils.py]: Don't use parenthesis `()` around the `if` clause.

File: BO_training/bo/localbo_utils.py [hunk 2]
```diff
@@ -407,14 +407,22 @@ def interleaved_search(x_center, f: Callable,
...
+        x0 = x0_cont
+    else:
+        x0_cat = np.array([deepcopy(random_sample_within_discrete_tr_ordinal(x_center[cat_dims], max_hamming_dist, config))
+                       for _ in range(n_restart)])
```
Comment (gmarcais) [inline BO_training/bo/localbo_utils.py]: Align the `for` with the open bracket `[`

File: BO_training/bo/localbo_utils.py [hunk 3]
```diff
@@ -407,14 +407,22 @@ def interleaved_search(x_center, f: Callable,
 
     # select the initialising points for both the continuous and categorical variables and then hstack them together
     # x0_cat = np.array([deepcopy(sample_neighbour_ordinal(x_center[cat_dims], config)) for _ in range(n_restart)])
-    x0_cat = np.array([deepcopy(random_sample_within_discrete_tr_ordinal(x_center[cat_dims], max_hamming_dist, config))
-                       for _ in range(n_restart)])
+    #x0_cat = np.array([deepcopy(random_sample_within_discrete_tr_ordinal(x_center[cat_dims], max_hamming_dist, config))
+    #                   for _ in range(n_restart)])
     # x0_cat = np.array([deepcopy(x_center[cat_dims]) for _ in range(n_restart)])
```
Comment (gmarcais) [inline BO_training/bo/localbo_utils.py]: Remove above 3 commented lines. They don't help the understanding of the code.

File: BO_training/ca_warmup_yaml.pl [hunk 4]
```diff
@@ -197,6 +197,7 @@ sub run_with_one_change{
       #}
 
       # Compute AUC
+      #TODO: change hard code
```
Comment (gmarcais) [inline BO_training/ca_warmup_yaml.pl:180]: What does `change hard code` mean?

File: contrastive_training/contrastive_main.py [hunk 5]
```diff
@@ -49,23 +49,23 @@
 #input_mash_mat = torch.from_numpy(input_mash_mat_np).to(dtype=dtype,device=device)
 #data_size = input_mash_mat.shape[0]
 
-input_mash_mat_subsample_np = np.load("/contrastive_training/input_features.npy")
+input_mash_mat_subsample_np = np.load("C:/Users/Zhiwen Yan/Desktop/23summer/input_features.npy")
```
Comment (gmarcais) [inline contrastive_training/contrastive_main.py:52]: Can't have absolute paths like this one. It either need to be a parameter, or a relative path that will work every time.

File: contrastive_training/contrastive_main.py [hunk 6]
```diff
@@ -49,23 +49,23 @@
...
 input_mash_mat_subsample = torch.from_numpy(input_mash_mat_subsample_np).to(dtype=dtype,device=device)
 
 input_mash_mat_np = input_mash_mat_subsample_np[np.arange(training_size)*8+7]
 input_mash_mat = torch.from_numpy(input_mash_mat_np).to(dtype=dtype,device=device)
 data_size = input_mash_mat.shape[0]
 
 #similarity matrix
-sim_mat_np = np.load("similarity_mat.npy")
+sim_mat_np = np.load("C:/Users/Zhiwen Yan/Desktop/23summer/similarity_mat.npy")
```
Comment (gmarcais) [inline contrastive_training/contrastive_main.py:60]: Hard coded path.

File: contrastive_training/contrastive_main.py [hunk 7]
```diff
@@ -49,23 +49,23 @@
...
 #data_size = sim_mat_np.shape[0]
 sim_mat_np = (np.ones((data_size,data_size))-np.eye(data_size))*sim_mat_np
 sim_mat_np = sim_mat_np/sim_mat_np.max()
 sim_mat = torch.from_numpy(sim_mat_np).to(dtype=dtype,device=device)
 assert data_size == sim_mat.shape[0]
 
 #similarity matrix with subsampling
-sim_mat_subsample_np = np.load("similarity_mat_subsample.npy")
+sim_mat_subsample_np = np.load("C:/Users/Zhiwen Yan/Desktop/23summer/similarity_mat_subsample.npy")
```
Comment (gmarcais) [inline contrastive_training/contrastive_main.py:68]: Hard coded path.

File: contrastive_training/contrastive_main.py [hunk 8]
```diff
@@ -128,11 +129,12 @@
 #training loop
 #pdb.set_trace()
 if use_subsample==1:
-    cmd = "mkdir -p ./contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)+"_subsample"
+    cmd = "mkdir -p .\contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)+"_subsample"
```
Comment (gmarcais) [inline contrastive_training/contrastive_main.py:132]: Please run things on the Ocean machines. Do not add Windows' specific notation in the code.

Ideally the code could work on any platform, but Windows is not a requirement. It is not used on the back-end (server, cloud, etc.). Your code MUST work for UNIX/Linux.

File: contrastive_training/contrastive_main.py [hunk 9]
```diff
@@ -128,11 +129,12 @@
...
     output_path = "./contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)+"_subsample/"
 else:
-    cmd = "mkdir -p ./contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)
+    cmd = "mkdir -p .\contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)
     output_path = "./contrastive_vali_figures_"+str(training_size)+"_lmda_"+lmda_str+"_beta_"+beta_str+"_mode_"+args.mode+"_temp_"+str(args.temperature)+"_wdecay_"+str(args.weightdecay)+"/"
+print(cmd)
 os.system(cmd)
```
Comment (gmarcais) [inline contrastive_training/contrastive_main.py:138]: Don't use `os.system("mkdir blah/blah/blah")` but rather `os.makedirs("blah/blah/blah")`. More likely to be portable, less error prone and faster.

See https://docs.python.org/3/library/os.html

File: contrastive_training/test.py [hunk 10]
```diff
@@ -1,3 +1,3 @@
 import numpy as np 
 
-data = np.load('input_features.npy')
\ No newline at end of file
+data = np.load('C:/Users/Zhiwen Yan/Desktop/23summer/input_features.npy')
```
Comment (gmarcais) [inline contrastive_training/test.py:3]: Hard coded path.

File: setting.yml [hunk 11]
```diff
@@ -24,37 +24,65 @@ main_args:
...
+
 #initial command of assemble software
-scallop:
-  software: scallop
-  input_command: ' -i '
-  additional_command: ' --verbose 0 --min_transcript_coverage 0'
+initial_command:
+  scallop:
+    name: scallop
+    input_command: ' -i '
```
Comment (gmarcais) [inline setting.yml:9]: These are not "command" but "switch" or "option"

**Review by gmarcais — CHANGES_REQUESTED** [review]
We should have a discussion regarding the format of the configuration YAML file.

File: contrastive_training/contrastive_main.py (hunk 5 above)
Comment (Yanzhiwen001) [reply contrastive_training/contrastive_main.py:52]: Contrastive part has not been generalized, just a test version on local machine.

File: contrastive_training/contrastive_main.py (hunk 7 above)
Comment (Yanzhiwen001) [reply contrastive_training/contrastive_main.py:68]: Contrastive part has not been generalized, just a test version on local machine.

File: contrastive_training/contrastive_main.py (hunk 8 above)
Comment (Yanzhiwen001) [reply contrastive_training/contrastive_main.py:132]: Contrastive part has not been generalized, just a test version on local machine.

File: BO_training/bo/optimizer_mixed.py [hunk 12]
```diff
@@ -154,8 +154,14 @@ def restart(self):
                 )
                 X_init_cat.append(cat_sample)
             X_init_cat = np.array(X_init_cat)
-            self.X_init = np.hstack((X_init_cat, X_init_cont))
-
+            #DONE: change stack manar into insert according to dim of cat
+            #self.X_init = np.hstack((X_init_cat, X_init_cont))
+            col = 0
+            for i in self.d_cat:
```
Comment (gmarcais) [inline BO_training/bo/optimizer_mixed.py:160]: Use `for col, i in enumerate(self.d_cat):` and remove `col = 0` and `col += 1` lines. Simpler, shorter, and more explanatory.

File: BO_training/bo/optimizer_mixed.py [hunk 13]
```diff
@@ -173,7 +179,14 @@ def restart(self):
                 X_init_cat = self.warp_discrete(X_init_cat, )
             X_init_cat = onehot2ordinal(X_init_cat, self.cat_dims)
             # Put the two parts back by a hstack
-            self.X_init = np.hstack((X_init_cat, X_init_cont))
+            #DONE: change stack manar into insert according to dim of cat 
+            #self.X_init = np.hstack((X_init_cat, X_init_cont))
+            col = 0
+            for i in self.d_cat:
```
Comment (gmarcais) [inline BO_training/bo/optimizer_mixed.py:185]: Same: `enumerate()`.

File: BO_training/ca_warmup_yaml.pl [hunk 14]
```diff
@@ -4,6 +4,10 @@
 use threads;
 use threads::shared;
 use Data::Dumper;
+#use YAML qw(LoadFile);
```
Comment (gmarcais) [inline BO_training/ca_warmup_yaml.pl:7]: Remove comment.

**Review by gmarcais — APPROVED** [review]
The Perl file in particular still seem to have some hard coded stuff. Bt I presume that what is working on next.

**Pull request merged**